target_chains/ton/contracts/contracts/Pyth.fc

```diff
@@ -176,29 +168,217 @@ int parse_pyth_payload_in_wormhole_vm(slice payload) impure {
 
     ;; Check if the sender has sent enough TON to cover the update_fee
     throw_unless(ERROR_INSUFFICIENT_FEE, remaining_msg_value >= update_fee);
+}
+
+tuple parse_price_feeds_from_data(int msg_value, slice data, tuple price_ids, int min_publish_time, int max_publish_time, int unique) {
+    slice cs = read_and_verify_header(data);
+
+    int wormhole_proof_size_bytes = cs~load_uint(16);
+    (cell wormhole_proof, slice new_cs) = read_and_store_large_data(cs, wormhole_proof_size_bytes * 8);
+    cs = new_cs;
+
+    int num_updates = cs~load_uint(8);
+
+    calculate_and_validate_fees(msg_value, num_updates);
 
     (_, _, _, _, int emitter_chain_id, int emitter_address, _, _, slice payload, _) = parse_and_verify_wormhole_vm(wormhole_proof.begin_parse());
 
     ;; Check if the data source is valid
     cell data_source = begin_cell()
         .store_uint(emitter_chain_id, 16)
         .store_uint(emitter_address, 256)
-        .end_cell();
+    .end_cell();
 
     ;; Dictionary doesn't support cell as key, so we use cell_hash to create a 256-bit integer key
     int data_source_key = cell_hash(data_source);
-
     (slice value, int found?) = is_valid_data_source.udict_get?(256, data_source_key);
     throw_unless(ERROR_UPDATE_DATA_SOURCE_NOT_FOUND, found?);
     int valid = value~load_int(1);
     throw_unless(ERROR_INVALID_UPDATE_DATA_SOURCE, valid);
 
-
     int root_digest = parse_pyth_payload_in_wormhole_vm(payload);
 
+    ;; Create tuple to store price feeds
+    tuple price_feeds = empty_tuple();
+    ;; Track which price IDs we've found
+    cell found_price_ids = new_dict();
+
     repeat(num_updates) {
-        (int price_id, int price, int conf, int expo, int publish_time, _, int ema_price, int ema_conf, slice new_cs) = read_and_verify_message(cs, root_digest);
+        (int price_id, int price, int conf, int expo, int publish_time, int prev_publish_time, int ema_price, int ema_conf, slice new_cs) = read_and_verify_message(cs, root_digest);
         cs = new_cs;
+        int should_include = (price_ids.tlen() == 0)
+            | ((price_ids.tlen() > 0)
+            & (publish_time >= min_publish_time)
+            & (publish_time <= max_publish_time)
+            & ((unique == 0) | (min_publish_time > prev_publish_time)));
+
+        if (should_include) {
+            ;; Create price feed cell containing both current and EMA prices
+            cell price_feed_cell = begin_cell()
+                .store_ref(store_price(price, conf, expo, publish_time))
+                .store_ref(store_price(ema_price, ema_conf, expo, publish_time))
+            .end_cell();
+
+            tuple price_feed = empty_tuple();
+            price_feed~tpush(price_id);
+            price_feed~tpush(price_feed_cell);
+            price_feeds~tpush(price_feed);
+
+            ;; Mark this price ID as found
+            found_price_ids~udict_set(256, price_id, begin_cell().store_int(true, 1).end_cell().begin_parse());
```

Comment:
if by any chance we have two updates for the same feed in a message we will add both of them to the result which is not desired.